src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```
@@ -561,11 +561,13 @@ Copyright (c) .NET Foundation. All rights reserved.
       <ResolvedFileToPublish Include="@(_SourceItemsToCopyToPublishDirectoryAlways)">
         <RelativePath>%(_SourceItemsToCopyToPublishDirectoryAlways.TargetPath)</RelativePath>
         <CopyToPublishDirectory>Always</CopyToPublishDirectory>
+        <IsKeyOutput Condition="@(_SourceItemsToCopyToPublishDirectoryAlways->'%(FullPath)') == $(AppHostIntermediatePath)">True</IsKeyOutput>
```

Feels weird that this isn't a % + .

Is IsKeyOutput a standard thing elsewhere?

---

Yes, this is a well established pattern.  CPS looks for this property in it's implementation of IVsOutputGroup2.get_KeyOutput.

Can you elaborate on the first part of that comment?  This is trying to set the metadata IsKeyOutput to true when the item's full path is equal to the value of the AppHostIntermediatePath property.  How exactly would you change it?

---

```suggestion
        <IsKeyOutput Condition="'%(_SourceItemsToCopyToPublishDirectoryAlways.FullPath)' == '$(AppHostIntermediatePath)'">True</IsKeyOutput>
```

@rainersigwald in case I'm missing something. Seems like we can access FullPath same as TargetPath above instead of transform over the single item in the batch. To me that's easier to read. I also added quotes for consistency, though I'm unsure of any semantic difference. I think it can come into play if LHS or RHS is not defined?

---

I also recommend quotes for consistency, the rules are too confusing to remember.

For `@(_SourceItemsToCopyToPublishDirectoryAlways->'%(FullPath)')` vs `%(_SourceItemsToCopyToPublishDirectoryAlways.FullPath)`, the batching engine will ensure that the item creation here is called once per unique set of `%(Item.Metadata)` values. If `FullPath` is not unique per `TargetPath`, the resulting batches could be confusing.

In that same condition before, `@(_SourceItemsToCopyToPublishDirectoryAlways->'%(FullPath)')` would be multi-valued and the condition would fail, so I don't think that's much worse here.

If `TargetPath` is unique (as we hope it would be to prevent double writes), there should be no meaningful difference between the two.

---

I've made this change and agree that it reads much better this way.  Thanks to you both.